Partially implement oauth

robotmayo opened the pull request:
This PR mostly address https://github.com/speedrun-website/leaderboard-backend/issues/10 

This PR lays down the basic foundation for OAUTH support. Right now it doesn't create or do anything with JWTs because there is currently no standard way to create/manipulate JWTs. I thought about adding that but this PR is already quite large so I wanted to cut scope wherever I could. As noted in the a few of the comments there are somethings that we could use such as a standard JSON out and Error response interface. Something like error codes(and types) would also be helpful at simplifying our error responses and make testing a bit better since we would no longer have to check the human readable strings.

Another noted issue is avoiding talking to gorm/database directly and use functions to fetch and write to the database. Another PR is working on something similar (https://github.com/speedrun-website/leaderboard-backend/pull/60). Im also not super great on the tests as its been a while since I have written production level Go advice in this area would be greatly appreciated.

robotmayo committed:
make user password nullable

robotmayo committed:
update password hash to be inside nullable string

robotmayo committed:
finish oauth skeleton

robotmayo committed:
change to user personal(temp)

robotmayo committed:
hopefully last cleanup pass

robotmayo committed:
minor routing changes, update comment

robotmayo committed:
remove uneeded param

robotmayo committed:
newline

robotmayo committed:
hopefullly the end

robotmayo committed:
update with twitter oauth variables

robotmayo committed:
add new line

robotmayo committed:
add new env vars

robotmayo committed:
add provider check

robotmayo committed:
add test for provider check

robotmayo committed:
tidy go sum

robotmayo committed:
remove template

BraydonKains commented on handlers/oauth/database.go:
```diff
@@ -0,0 +1,52 @@
+	if result.Error == nil {
+		return &maybeExistingUser, nil
+	}
+	isNotFoundError := errors.Is(result.Error, gorm.ErrRecordNotFound)
```
nit: As a style thing I personally prefer just doing `errors.Is` inline in the `if` condition instead of the transient variable. I think it reads slightly better. Not a requirement though if you feel differently.

BraydonKains commented on handlers/oauth/database.go:
```diff
@@ -0,0 +1,52 @@
+	if !isNotFoundError {
+		return nil, result.Error
+	}
+	return nil, nil
```
Generally if the function did was not successful, it should return some kind of error; returning `nil` for the error generally implies success. Why do we want to return `nil` in this case? Maybe there's a compromise that still returns some kind of error here.

robotmayo replied:
Im not entirely sure what our policy should be here. This is technically not an error as not find a user is perfectly expected. I think the only reason gorm errors is because it attempts to update a pointer and theres no easy to say that 'failed'.

zysim replied:
I'd say that in this function that we call `GetUserByTwitterID`, I'd expect _to_ find a user. And since we couldn't find a user with that Twitter ID, you could reasonably throw an error. Maybe we can have a `UserNotFoundError` which we return in both cases where Gorm errors (i.e. `gorm.ErrRecordNotFound`) and when a record simply doesn't exist.

BraydonKains commented on handlers/oauth/database.go:
```diff
@@ -0,0 +1,52 @@
+
+//CreateUser creates a user
+//
+//TODO: Replace with userstore method?
```
That would be good, but we can have a new PR that stitches our stuff together as expected rather than grow this one.

BraydonKains commented on handlers/oauth/oauth.go:
```diff
@@ -0,0 +1,142 @@
+	for _, provider := range enabledProviders {
+		cleanedProvider := strings.ToUpper(strings.TrimSpace(provider))
+		switch cleanedProvider {
+		case twitterProvider.Name():
```
This is fine, but something that may clean it up:
If we have a `map[string]*Provider` where the `string` keys are values we expect to get from the env var, then we can both avoid the string transformation we have to do here, and avoid a potentially growing `switch` statement.

BraydonKains commented on handlers/oauth/oauth.go:
```diff
@@ -0,0 +1,142 @@
+		c.JSON(http.StatusOK,
+			//We "copy" the struct here to ensure that response is consistent
+			//even if the type of the existing user changes
+			&model.UserIdentifier{
```
Note: This will be moved to using the `data` top level key added in #60, we can fix it in the PR that will align this work to that PR. So it should be fine not to block this PR on that for now.

BraydonKains commented on handlers/oauth/oauth_test.go:
```diff
@@ -0,0 +1,286 @@
+	return newUser, nil
+}
+
+func Test_OauthCallbackUserAuthError(t *testing.T) {
```
I'd prefer the test function names not to have underscores, and I think that's it's suggested to just name things as `TestName` (see https://golang.org/doc/tutorial/add-a-test)

zysim replied:
Got rid of them underscores in https://github.com/leaderboardsgg/leaderboard-backend/pull/69/commits/44de6ebe33d546c4f8ac172dc6bcae6c25bd6e7a

BraydonKains commented on handlers/oauth/oauth_test.go:
```diff
@@ -0,0 +1,286 @@
+package oauth
```
In the other PR I used the `_test` package pattern. We'll have to decide as a team which we prefer.

zysim replied:
As in the file names are `*_test.go`? If so then yeah sure; don't see a reason to not do that.

FrozenSake replied:
You can also have the package be `package oauth_test`, Some reading about this pattern [here](https://github.com/golang/go/issues/25223). I believe it's generally beneficial to do because it causes the testing package to have to import, which can catch some gotchas.

BraydonKains commented on handlers/oauth/oauth_test.go:
```diff
@@ -0,0 +1,286 @@
+	store := initStore()
+	expectedUser, createUserErr := store.CreateUser(model.User{
+		Email:    "oauthcallbackereturn@example.com",
+		Username: "iliketurtles",
```
Me too

BraydonKains commented on handlers/oauth/oauth_test.go:
```diff
@@ -0,0 +1,286 @@
+
+func Test_OauthCallbackReturnsExistingUser(t *testing.T) {
+	store := initStore()
+	expectedUser, createUserErr := store.CreateUser(model.User{
```
nit: Usually in a scenario like this where the error is immediately checked, I'd prefer just sticking to the naming convention of `err`. This comment applies to a few other places as well.

zysim replied:
Done with https://github.com/leaderboardsgg/leaderboard-backend/pull/69/commits/672925b928f6e568330b0f0dba62f5406e071d46

BraydonKains commented on handlers/oauth/oauth_test.go:
```diff
@@ -0,0 +1,286 @@
+}
+
+func Test_InitializeProviders(t *testing.T) {
+	setEnvErr := os.Setenv("ENABLED_PROVIDERS", "twitter")
```
nit: Similar to above. I would name these errors as you have here in a scenario where they are all evaluated at the same time. Since these errors are all being handled sequentially after being received, I'd prefer to stick to the convention of naming it `err`.

zysim replied:
Done with https://github.com/leaderboardsgg/leaderboard-backend/pull/69/commits/672925b928f6e568330b0f0dba62f5406e071d46

BraydonKains commented on middleware/jwt.go:
```diff
@@ -53,7 +53,12 @@ var JwtConfig = &jwt.GinJWTMiddleware{
+			log.Println("User password in database is null indicating they use oauth and not the password flow")
+			return nil, jwt.ErrFailedAuthentication
+		}
+		passwordMatches, _ := utils.ComparePasswords(user.Password, []byte(password))
```
Since we've started returning the error from `ComparePasswords` now, we might as well use it here. Even though we're pretty much certain it will always be `nil` in this case, if it's returning an error I think we should use it.

RageCage64 reviewed (changes requested):
Thanks for getting this started! This is one of the trickiest parts of the solution and there isn't much of a codebase to work with yet, so this can't have been a terribly straightforward PR to put together.

I left some "nit" comments, which are mostly pedantic style things, some questions/suggestions, and some small things I'd like to see changed.

robotmayo committed:
inline error.Is check

robotmayo committed:
clean up provider checking

robotmayo committed:
fix oauth test referencing twitter

robotmayo committed:
address error

robotmayo committed:
change package, fix a few references

zysim commented on handlers/oauth/oauth_test.go:
```diff
@@ -278,7 +278,7 @@ func Test_InitializeProviders(t *testing.T) {
 	}
 
-	provider, getProviderErr := goth.GetProvider(twitterProvider.Name())
+	provider, getProviderErr := goth.GetProvider("twitter")
```
Have we decided to just use strings here instead? That said, I don't mind this too much

zysim reviewed (commented):
~Looks good, I've just got that one nitpick is all.~
EDIT: Ignore me. Looks good to me; just need fixing conflicts and sorting out the other comments :+1:

Commit:
initial commit

Commit:
update authenticator to handle null

Commit:
scuffed twitter login

Commit:
final cleanup

Commit:
Tidying Up Force-pushing Boo-boos

Commit:
Fix Package Import and Failing Test

Commit:
Remove Underscores in Test Function Names

Commit:
Sticking to Simple Naming Convention
Our errors in oauth_test.go are immediately checked for now, so we're
sticking to calling em "err".
